Use API for description searches

rrotter opened the pull request:
Both `brew search --desc` and `brew desc --search` use API for cask and formula searches unless `--eval-all` or `HOMEBREW_EVAL_ALL` set. Description searches do not use the description cache or eval any formulas/casks.

With `--eval-all`, description search reverts to the old behavior.

This will resolve #16237.

- [x] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/HEAD/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [x] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/brew/blob/HEAD/Library/Homebrew/test/PATH_spec.rb).
- [x] Have you successfully run `brew style` with your changes locally?
- [x] Have you successfully run `brew typecheck` with your changes locally?
- [x] Have you successfully run `brew tests` with your changes locally?

-----

This is intentionally a draft, looking for comments on this direction.

MikeMcQuaid commented on Library/Homebrew/descriptions.rb:
```diff
@@ -9,7 +9,8 @@
 class Descriptions
   # Given a regex, find all formulae whose specified fields contain a match.
   def self.search(string_or_regex, field, cache_store, eval_all = Homebrew::EnvConfig.eval_all?)
```
```suggestion
  def self.search(string_or_regex, field, cache_store, eval_all = Homebrew::EnvConfig.eval_all?, cache_store_hash: false)
```

MikeMcQuaid commented on Library/Homebrew/descriptions.rb:
```diff
@@ -9,7 +9,8 @@
   # Given a regex, find all formulae whose specified fields contain a match.
   def self.search(string_or_regex, field, cache_store, eval_all = Homebrew::EnvConfig.eval_all?)
-    cache_store.populate_if_empty!(eval_all:)
+    # allow searching a simple hash, not only cache_store objects
+    cache_store.populate_if_empty!(eval_all:) if cache_store.methods.include?(:populate_if_empty!)
```
```suggestion
    cache_store.populate_if_empty!(eval_all:) unless cache_store_hash
```

MikeMcQuaid commented on Library/Homebrew/search.rb:
```diff
@@ -22,19 +22,29 @@ def self.search_descriptions(string_or_regex, args, search_type: :desc)
+          end
+        else
+          descriptions = Homebrew::API::Formula.all_formulae.transform_values { |data| data["desc"] }
+          Descriptions.search(string_or_regex, search_type, descriptions, eval_all).print
```
```suggestion
          Descriptions.search(string_or_regex, search_type, descriptions, eval_all, cache_store_hash: true).print
```

MikeMcQuaid commented on Library/Homebrew/search.rb:
```diff
@@ -22,19 +22,29 @@ def self.search_descriptions(string_or_regex, args, search_type: :desc)
+        end
+      else
+        descriptions = Homebrew::API::Cask.all_casks.transform_values { |data| data["desc"] }
+        Descriptions.search(string_or_regex, search_type, descriptions, eval_all).print
```
```suggestion
        Descriptions.search(string_or_regex, search_type, descriptions, eval_all, cache_store_hash: true).print
```

MikeMcQuaid commented on Library/Homebrew/search.rb:
```diff
@@ -22,19 +22,29 @@ def self.search_descriptions(string_or_regex, args, search_type: :desc)
+            Descriptions.search(string_or_regex, search_type, cache_store, eval_all).print
+          end
+        else
+          descriptions = Homebrew::API::Formula.all_formulae.transform_values { |data| data["desc"] }
```
I wonder if this warrants printing a warning if there's non-core taps with formulae tapped (as they won't be checked)?

rrotter replied:
I think so. Was expecting to add this, just wanted to get the core logic committed first. Current code has warnings for both formulae and casks, and they only print if there are 3rd party taps installed that include formulae/casks.

MikeMcQuaid commented on Library/Homebrew/search.rb:
```diff
@@ -22,19 +22,29 @@ def self.search_descriptions(string_or_regex, args, search_type: :desc)
+          Descriptions.search(string_or_regex, search_type, cache_store, eval_all).print
+        end
+      else
+        descriptions = Homebrew::API::Cask.all_casks.transform_values { |data| data["desc"] }
```
I wonder if this warrants printing a warning if there's non-homebrew-cask taps with casks tapped (as they won't be checked)?

rrotter replied:
^see comment on formulae.

MikeMcQuaid reviewed (commented):
This is looking great so far, nice work @rrotter!

Assuming you agree with suggestions: only thing I think is needed is to get some test coverage and to decide how to message the non-eval-all case if there's other taps.

apainintheneck commented:
Does this change the behavior of cask description searches? I remember previously that it was using both the names and descriptions in search and it looks like it's now only using descriptions. This is useful when people want to surface cask results for alternative names.

apainintheneck commented:
It'd also be nice to get some benchmarks at some point too.

rrotter commented:
> It'd also be nice to get some benchmarks at some point too.

In my testing this search takes about 1s, and the old method was 0.5s cached and 5s uncached.

rrotter commented:
> Does this change the behavior of cask description searches?

Good catch. This is using exactly the same code to perform a search, but I rewrote the code that generates the hash that is being searched. I didn't realize that the data structure was slightly different for formulae and casks. The cask data structure actually has a third field.
```
# formula description hash
{"name"=>"description"}
# cask description hash
{"token"=>["list, of, names", "description"]}
```
Fixed in my latest commit.

rrotter commented:
Here's the current warning language that I'm printing if there are third party taps installed. Can't paste w/ color here, but I'm using `opoo`.
```
~ % brew desc -s foo
==> Formulae
Warning: Use `--eval-all` to search 6 additional formulae in third party taps.
chibi-scheme: Small footprint Scheme for use as a C Extension Language
duktape: Embeddable Javascript engine with compact footprint
foremost: Console program to recover files based on their headers and footers
...
```

Trying to keep it clear and concise. Could add something about why you might not want to use `--eval-all`, but that would get long. Currently just under 80 chars (obv length varies depending upon number of external formulae).

rrotter commented:
I'm a little stuck on testing:
* The one test I added (checks that `search --desc` works w/o `--eval-all`) just checks that the search completes w/o failure, but doesn't check the search results because, AFAIK, we don't have a way to mock the API to include "testball". Maybe this is the best I can do for now? (probably relates to #16895)
* I'd like to test that the warning is printed or not printed as appropriate. I think the way to do this is to add a test tap w/ a test formula in it, but I can't quite get this to work (and I don't see any similar examples in the existing test code). Here's what I've got so far:
```
tap = Tap.from_path(setup_test_tap)
setup_test_formula "testballexternal", tap: tap
```
```
Failure/Error: setup_test_formula "testballexternal", tap: tap

Errno::ENOENT:
   No such file or directory @ rb_sysopen - /private/tmp/homebrew-tests-20240629-39807-bbynmd/prefix/Library/Taps/homebrew/homebrew-foo/Formula/testballexternal.rb
```

Any tips would be appreciated. Other than adding tests I should be done pushing on this branch for now.

apainintheneck commented on Library/Homebrew/cmd/desc.rb:
```diff
@@ -47,10 +47,6 @@ def run
         end
 
         if search_type.present?
-          if !args.eval_all? && !Homebrew::EnvConfig.eval_all?
-            raise UsageError, "`brew desc --search` needs `--eval-all` passed or `HOMEBREW_EVAL_ALL` set!"
-          end
```
Suggestion: We still probably want to keep this error message when `HOMEBREW_NO_INSTALL_FROM_API` is set. It's possible that no API JSON file exists or that it's outdated which would cause unexpected results. This is admittedly a rare case though.

apainintheneck replied:
You can use `Homebrew::EnvConfig.no_install_from_api?` to check that environment variable.

rrotter replied:
done.

apainintheneck commented on Library/Homebrew/cmd/search.rb:
```diff
@@ -70,10 +70,6 @@ def run
         string_or_regex = Search.query_regexp(query)
 
         if args.desc?
-          if !args.eval_all? && !Homebrew::EnvConfig.eval_all?
-            raise UsageError, "`brew search --desc` needs `--eval-all` passed or `HOMEBREW_EVAL_ALL` set!"
-          end
```
Ditto https://github.com/Homebrew/brew/pull/17582#discussion_r1660064348

apainintheneck commented:
The performance looks good and adding cask names was a simple fix which was nice.

> I'm a little stuck on testing:
> 
> * The one test I added (checks that `search --desc` works w/o `--eval-all`) just checks that the search completes w/o failure, but doesn't check the search results because, AFAIK, we don't have a way to mock the API to include "testball". Maybe this is the best I can do for now? (probably relates to [Run the test suite in the default API mode #16895](https://github.com/Homebrew/brew/issues/16895))
> * I'd like to test that the warning is printed or not printed as appropriate. I think the way to do this is to add a test tap w/ a test formula in it, but I can't quite get this to work (and I don't see any similar examples in the existing test code). Here's what I've got so far:
> 
> ```
> tap = Tap.from_path(setup_test_tap)
> setup_test_formula "testballexternal", tap: tap
> ```
> 
>  ```
> Failure/Error: setup_test_formula "testballexternal", tap: tap
> 
> Errno::ENOENT:
>    No such file or directory @ rb_sysopen - /private/tmp/homebrew-tests-20240629-39807-bbynmd/prefix/Library/Taps/homebrew/homebrew-foo/Formula/testballexternal.rb
> ```
> 
> Any tips would be appreciated. Other than adding tests I should be done pushing on this branch for now.

In terms of integration tests, nobody's been able to find a good way to mock and sign the API JSON yet. I think it'd make more sense to add unit tests for `Homebrew::Search.search_descriptions` instead. It would likely require mocking the API though which we do already in a few places (see . Certainly doable. That being said unit tests would be nice to have but aren't a blocker here since the surface area of the change is pretty small.

- Some API unit tests for reference
https://github.com/Homebrew/brew/blob/fda555d7e61d13a142f1505dc58591add3c09d29/Library/Homebrew/test/formulary_spec.rb#L271
https://github.com/Homebrew/brew/blob/fda555d7e61d13a142f1505dc58591add3c09d29/Library/Homebrew/test/cask/cask_loader/from_api_loader_spec.rb#L3

rrotter commented on Library/Homebrew/test/search_spec.rb:
```diff
@@ -57,4 +58,33 @@
   end
+
+  describe "#search_descriptions" do
+    let(:args) { instance_double("args") }
```
This line is failing `brew style`. `--fix` just removes the quotes which completely breaks the code. There should be an additional argument here of what class `args` is supposed to be, but that just gives a different error:
```
the Homebrew::CLI::Args class does not implement the instance method: formula?
the Homebrew::Cmd::Desc::Args class does not implement the instance method: formula?
```

apainintheneck replied:
I was going to suggest using `OpenStruct` but it looks like we have a linter rule against that too. This works but is kind of hacky.

```diff
diff --git a/Library/Homebrew/test/search_spec.rb b/Library/Homebrew/test/search_spec.rb
index 11f5d2c9dc..afb409f2a9 100644
--- a/Library/Homebrew/test/search_spec.rb
+++ b/Library/Homebrew/test/search_spec.rb
@@ -2,6 +2,7 @@
 
 require "search"
 require "descriptions"
+require "cmd/desc"
 
 RSpec.describe Homebrew::Search do
   describe "#query_regexp" do
@@ -60,7 +61,7 @@ RSpec.describe Homebrew::Search do
   end
 
   describe "#search_descriptions" do
-    let(:args) { instance_double("args") }
+    let(:args) { Homebrew::Cmd::Desc.new(["min_arg_placeholder"]).args }
 
     context "with api" do
       def api_formulae
@@ -72,7 +73,6 @@ RSpec.describe Homebrew::Search do
       end
 
       before do
-        allow(args).to receive_messages(formula?: false, cask?: false, eval_all?: false)
         allow(Homebrew::API::Formula).to receive(:all_formulae).and_return(api_formulae)
         allow(Homebrew::API::Cask).to receive(:all_casks).and_return(api_casks)
       end
```

@dduugg Do you have any better ideas here?

rrotter replied:
I committed this, since it works and I don't need to mock multiple argument combinations.

dduugg replied:
I thought you could use KV pairs to implement instance methods, e.g. `let(:args) { instance_double(Homebrew::CLI::Args, "args", formula?: true) }` or something.

OpenStruct is a bad idea even in tests because it will fail open, e.g. `OpenStruct.new.not_a_method`, while `instance_double("args").not_a_method` will. 

(Apologies if i haven't captured the question, i see this went in, and thus I didn't spend much time gathering context.)

apainintheneck replied:
Yeah, that probably would have worked too. Good point.

rrotter commented:
Pushed a couple of tests, including an API mock. They do pass, but CI is failing on the linter, see comment above.

Other than the lint, does this look okay?

apainintheneck commented:
> Pushed a couple of tests, including an API mock. They do pass, but CI is failing on the linter, see comment above.
> 
> Other than the lint, does this look okay?

@rrotter Yep, the code looks good. I left an optional nit as well but feel free to ignore that.

rrotter commented:
Thanks, @apainintheneck, added your suggestions, squashed and rebased.

apainintheneck reviewed (approved):
LGTM, thanks for fixing this!

MikeMcQuaid reviewed (approved):
Thanks @rrotter!
